### DISCUSSION TITLE
How to save the data when the tonic is closed!

### DISCUSSION THREAD

**Discussion by xiuno** [hyperium/tonic#725] [Q&A] [answered]
The user declared variable is moved to the structure, and the Route does not provide a way to access it, is there any good way ? 

```
#[tokio::main]
async fn main() {

    let game = Game { data: Arc::new(Mutex::new(vec![])) };
    let mut route = Server::builder().add_service(GameServer::new(game));

    // register callback when shutdown
    route.serve_with_shutdown(addr, async {
        tokio::signal::ctrl_c().await.unwrap();

        // how to fetch "game.data" in here ?
        // "game" already moved to GameServer
        // "GameServer" moved to Route again
        // "Route" doesn't provide a way to access "game"?
        let data = route.game.data; // <-- how to write ?

    }).await.unwrap();

}
```

I found following code in tonic/src/transport/server/mod.rs
```
/// A stack based `Service` router.
#[derive(Debug)]
pub struct Router<A, B> {
    server: Server,
    routes: Routes<A, B, Request<Body>>, // if this field is pub, will ok? or provide a method to visit for user data ?
}
```

**Comment by davidpdrsn** [comment, marked as answer]
The data is in an `Arc` so you can share it between the `GameServer` and whatever code runs after the server has been shutdown.

I haven't tested this code but should work

```rust
#[tokio::main]
async fn main() {
    // pull this out into a variable so it can be shared between tasks
    let data = Arc::new(Mutex::new(vec![]));

    let game = Game { data: Arc::clone(&data) };
    let mut route = Server::builder().add_service(GameServer::new(game));

    // register callback when shutdown
    route.serve_with_shutdown(addr, async move {
        tokio::signal::ctrl_c().await.unwrap();
    }).await.unwrap();

    // somehow actually save the data
    let data = data.lock();
    println!("saving {:?}", data);
}
```

**Reply by davidpdrsn** [reply]
Making `routes` public as you suggested wouldn't work since `serve_with_shutdown` takes `self` by value, so the routes aren't accessible afterwards.

**Reply by xiuno** [reply]
it works fine by your way, thanks!
But I still have some doubts. Why don't provide access from outside to inside? This may be more convenient to use.

**Reply by davidpdrsn** [reply]
The server requires owning the data for several reasons, so `serve_with_shutdown` couldn't take `&self`.

**Reply by xiuno** [reply]
got it

**Reply by xiuno** [reply]
I have encounted a new issue, if client don't disconnect positively(client build with the stream type, receive the push message from server),  service will block in .await forever even after receive the Ctrl+ C signal, how to disconnect the all connection? 
 
 in my.proto
 ```
 service Xxxx {
   ...
   rpc PushMessage (EmptyRequest) returns (stream MsgResponse) {}
}
  ```
 
 #[tokio::main]
async fn main() {
    // pull this out into a variable so it can be shared between tasks
    let data = Arc::new(Mutex::new(vec![]));

    let game = Game { data: Arc::clone(&data) };
    let mut route = Server::builder().add_service(GameServer::new(game));

    // register callback when shutdown
    route.serve_with_shutdown(addr, async move {
        tokio::signal::ctrl_c().await.unwrap();
        // println!("received ctrl+C"); // <--- can running to here.
        // tonic::alose_all_stream();   // how to close all client ?
    }).await.unwrap();

    // <---- can't reach here 
    // somehow actually save the data
    let data = data.lock();
    println!("saving {:?}", data);
}
